lib/google/ads/google_ads/deprecation.rb

```diff
@@ -0,0 +1,100 @@
+module Google
+  module Ads
+    module GoogleAds
+      class Deprecation
+        Error = Class.new(StandardError)
+        def initialize(
+          treat_deprecation_warnings_as_errors,
+          warn_on_all_deprecations
+        )
+          @treat_deprecation_warnings_as_errors = treat_deprecation_warnings_as_errors
+          @warn_on_all_deprecations = warn_on_all_deprecations
+          @silenced_sites = {}
+        end
+
+        def deprecate(deprecation)
+          return unless should_warn?
+          if @treat_deprecation_warnings_as_errors
+            raise Error, deprecation
+          else
+            Warning.warn("#{deprecation}. Called from: #{CallerFilter.first_non_google_ads_line}")
+          end
+        end
+
+        private
+
+        # Determines if we should issue a deprecation warning, silencing
+        # on each user call site after the first warning.
+        def should_warn?
+          return true if @warn_on_all_deprecations
+
+          if !@silenced_sites.include?(CallerFilter.first_non_google_ads_line)
+            @silenced_sites[CallerFilter.first_non_google_ads_line] ||= true
+            true
+          else
+            false
+          end
+        end
+      end
+
+      # @api private
+      # Portions of this implementation were copied from RSpec, under the
+      # MIT license: https://git.io/fjR7i
```

Comment:
@AnashOommen PTAL

Reply:
I *think* the use of this code is fine, but check with opensource-discuss@ before committing. I suspect that this code has to live in its own file, in a third_party folder instead of mixing it into your code.

Just curious, is there a ruby gem we can depend on instead for this snippet? That might be simpler.

Reply:
We're ok to use MIT licensed code according to opensource licensing.